Add support for `T & readonly`

MaryamZi opened the pull request:
## Purpose
$title, this allows creating values known to be immutable at compile-time.
Main changes to fix https://github.com/ballerina-platform/ballerina-lang/issues/22515.

Fixes #23050 - Rewrite mapping/list value creation

This PR also fixes #22950 - Setting members to a tuple assigned to a JSON array-typed variable does not work

## Samples
```ballerina
import ballerina/io;

type Employee record {|
    PersonalDetails details;
    string department;
|};

type PersonalDetails record {|
    string name;
    int id;
|};

public function main() {
    Employee e1 = {
        details: {
            name: "Jo",
            id: 1234	
        },
        department: "finance"
    };

    // Compilation Error: 
    // incompatible types: expected 'Employee & readonly', found 'Employee'
    // Employee & readonly e2 = e1; 
   
    // Immutable `Employee` value.
    Employee & readonly e3 = {
        details: {
            name: "May",
            id: 7890	
        },
        department: "finance"
    };   

    // Compilation Error: 
    // cannot update 'readonly' record field 'department' in 'Employee & readonly'
    // e3.department = "IT";

    io:println(e3.isReadOnly()); // true
}
```

## Remarks
To Do
- Tests (WIP)
- Allow constructing values with their read-only bit on with a type-cast-expr `<readonly>` #22954
- Disallow langlib methods that modify the value on immutable values #22964


## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

MaryamZi commented:
Closing this PR temporarily since some significant changes need to be introduced to improve how the immutable values are created. Will re-open once done.

Kishanthan commented:
@MaryamZi can you check for alignment issues in this PR?

MaryamZi commented:
> @MaryamZi can you check for alignment issues in this PR?

@Kishanthan do you mean in Java code or Ballerina? I'm not sure I understood where the issues are. Can you please point out one maybe?

manuranga commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmValueGen.java:
```diff
@@ -680,8 +688,12 @@ private void createRecordInitWrapper(ClassWriter cw, String className, BIRNode.B
+
                 mv.visitMethodInsn(INVOKESTATIC, refTypeClassName, "$init",
-                        String.format("(L%s;L%s;)V", STRAND, MAP_VALUE), false);
+                        String.format("(L%s;L%s;[L%s;)V", STRAND, MAP_VALUE, BINITIAL_VALUE_ENTRY), false);
```
I am in the process of removing `createRecordInitWrapper` method, can we please move (or duplicate) this in `createInstantiateMethod`

MaryamZi replied:
We don't need initial values for referenced record types, so having the initial value population logic in "instantiate" should be enough. Moved to `createInstantiateMethod` in 6f89a47.

Kishanthan commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:
```diff
@@ -1336,33 +1390,27 @@ void generateArrayStoreIns(BIRNonTerminator.FieldAccess inst) {
-            this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, "add", "(JB)V", true);
-        } else {
-            this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, "add", String.format("(JL%s;)V", OBJECT), true);
+            BType valueType = inst.rhsOp.variableDcl.type;
```
looks like there is an alignment issue here. probably due to merge conflicts with jvm codegen package classes.

Kishanthan commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:
```diff
@@ -1336,33 +1390,27 @@ void generateArrayStoreIns(BIRNonTerminator.FieldAccess inst) {
-            this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, "add", "(JB)V", true);
-        } else {
-            this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, "add", String.format("(JL%s;)V", OBJECT), true);
+            BType valueType = inst.rhsOp.variableDcl.type;
+
+            String method = "add";
+
+            if (TypeTags.isIntegerTypeTag(valueType.tag)) {
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method, "(JJ)V", true);
+            } else if (valueType.tag == TypeTags.FLOAT) {
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method, "(JD)V", true);
+            } else if (TypeTags.isStringTypeTag(valueType.tag)) {
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method,
+                                        String.format("(JL%s;)V", JvmPackageGen.IS_BSTRING ?
+                                                JvmConstants.B_STRING_VALUE : STRING_VALUE), true);
+            } else if (valueType.tag == TypeTags.BOOLEAN) {
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method, "(JZ)V", true);
+            } else if (valueType.tag == TypeTags.BYTE) {
+                this.mv.visitInsn(I2B);
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method, "(JB)V", true);
+            } else {
+                this.mv.visitMethodInsn(INVOKEINTERFACE, ARRAY_VALUE, method, String.format("(JL%s;)V", OBJECT), true);
+            }
```
looks like there is an alignment issue here. probably due to merge conflicts with jvm codegen package classes.

Kishanthan commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:
```diff
@@ -1583,14 +1631,17 @@ void generateFPLoadIns(BIRNonTerminator.FPLoad inst, LambdaMetadata lambdaMetada
-                        JvmPackageGen.IS_BSTRING ? JvmConstants.B_STRING_VALUE : STRING_VALUE, XML_VALUE), false);
-        this.storeToVar(newXMLElement.lhsOp.variableDcl);
-    }
+            this.loadVar(newXMLElement.startTagOp.variableDcl);
+            this.mv.visitTypeInsn(CHECKCAST, XML_QNAME);
+            this.loadVar(newXMLElement.defaultNsURIOp.variableDcl);
+            loadType(this.mv, newXMLElement.type);
+
+            this.mv.visitMethodInsn(INVOKESTATIC, XML_FACTORY, "createXMLElement",
+                                    String.format("(L%s;L%s;L%s;)L%s;", XML_QNAME,
+                                                  JvmPackageGen.IS_BSTRING ? JvmConstants.B_STRING_VALUE :
+                                                          STRING_VALUE, BTYPE, XML_VALUE), false);
+            this.storeToVar(newXMLElement.lhsOp.variableDcl);
+        }
```
looks like there is an alignment issue here. probably due to merge conflicts with jvm codegen package classes.

Kishanthan commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:
```diff
@@ -1627,23 +1678,28 @@ void generateNewXMLTextIns(BIRNonTerminator.NewXMLText newXMLText) {
-                String.format("(L%s;L%s;)L%s;", consVal, consVal, XML_VALUE), false);
-        this.storeToVar(newXMLPI.lhsOp.variableDcl);
-    }
+            this.loadVar(newXMLPI.targetOp.variableDcl);
+            this.loadVar(newXMLPI.dataOp.variableDcl);
+            loadType(this.mv, newXMLPI.type);
+
+            String consVal = JvmPackageGen.IS_BSTRING ? JvmConstants.B_STRING_VALUE : STRING_VALUE;
+            this.mv.visitMethodInsn(INVOKESTATIC, XML_FACTORY, "createXMLProcessingInstruction",
+                                    String.format("(L%s;L%s;L%s;)L%s;", consVal, consVal, BTYPE, XML_VALUE), false);
+            this.storeToVar(newXMLPI.lhsOp.variableDcl);
+        }
```
looks like there is an alignment issue here. probably due to merge conflicts with jvm codegen package classes.

Kishanthan commented:
> > @MaryamZi can you check for alignment issues in this PR?
> 
> @Kishanthan do you mean in Java code or Ballerina? I'm not sure I understood where the issues are. Can you please point out one maybe?

Pl check now. I have added few comments. you have to check the merge conflicts classes with jvm codegen package for this issue.

MaryamZi commented:
> > > @MaryamZi can you check for alignment issues in this PR?
> > 
> > 
> > @Kishanthan do you mean in Java code or Ballerina? I'm not sure I understood where the issues are. Can you please point out one maybe?
> 
> Pl check now. I have added few comments. you have to check the merge conflicts classes with jvm codegen package for this issue.

Thank you, missed this. Fixed in 6f89a47.

KRVPerera commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -2106,6 +2112,9 @@ public void visit(BLangFieldBasedAccess fieldAccessExpr) {
                     if (isAllReadonlyTypes(varRefType)) {
                         dlog.error(fieldAccessExpr.pos, DiagnosticCode.CANNOT_UPDATE_READONLY_VALUE_OF_TYPE,
                                    varRefType);
+                        types.checkType(fieldAccessExpr, actualType, this.expType);
```
Why do we call `checkType` here?

KRVPerera commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:
```diff
@@ -1334,7 +1334,8 @@ private void defineErrorDetails(List<BLangTypeDefinition> typeDefNodes, SymbolEn
 
     private void defineFields(List<BLangTypeDefinition> typeDefNodes, SymbolEnv pkgEnv) {
-        for (BLangTypeDefinition typeDef : typeDefNodes) {
+        for (int i = 0; i < typeDefNodes.size(); i++) {
```
Is this a performance improvement? I think this is unnecessary.

MaryamZi replied:
This is to avoid potential concurrent modification errors since we may have to define new record type defs while iterating through the current list.

MaryamZi commented on bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java:
```diff
@@ -1154,6 +1166,87 @@ private static boolean isReadonlyType(BType sourceType) {
+        return isSelectivelyImmutableType(type, new HashSet<>());
+    }
+
+    public static boolean isSelectivelyImmutableType(BType type, Set<BType> unresolvedTypes) {
```
No, this does not check for the read-only bit.

Even though mappings in general are selectively immutable, some mappings can never be immutable. e.g., `map<future>`. This method checks if the specified type can be immutable. 

If it's still unclear maybe we can call this something like `isPossiblyImmutableSelectivelyImmutableType`, but given that the logic is quite obvious I'm not sure we need it.

KRVPerera replied:
Okay. I was thinking if someone wanted to write a method in future to check the read-only bit; then we may need to change the name at that time.

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -2168,14 +2208,19 @@ public void visit(BLangIndexBasedAccess indexBasedAccessExpr) {
-            String name = getConstFieldName(indexExpr);
-            if (allRecordsHaveRequiredReadonlyField(exprType, name)) {
+                (indexBasedAccessExpr.lhsVar || indexBasedAccessExpr.compoundAssignmentLhsVar)) {
+            if (isAllReadonlyTypes(exprType)) {
```
Check for optional fields.

MaryamZi replied:
This is not required since what we are checking here is something like

```
(Foo & readonly)|(Bar & readonly) x = {};
x["val"] = 1;
```

This is always invalid since the value itself is immutable and cannot be updated.

MaryamZi commented on bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/XMLFactory.java:
```diff
@@ -345,6 +386,31 @@ public static XMLValue createXMLProcessingInstruction(BString tartget, BString d
+     * @param type      The type for the PI
+     * @return XMLValue Processing instruction type XMLValue
+     */
+    public static XMLValue createXMLProcessingInstruction(BString target, BString data, BType type) {
```
Accept a flag instead and set relevant type internally.

pubudu91 commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1481,7 +1497,12 @@ private boolean validateRequiredAndReadonlyFields(BRecordType type,
                 }
-                dlog.error(typePosPair.pos, DiagnosticCode.INVALID_FIELD_FOR_READONLY_RECORD_FIELD, fieldName);
+
+                if (mutableRecordType && !Symbols.isFlagOn(field.type.flags, Flags.READONLY)) {
```
The `!Symbols.isFlagOn(field.type.flags, Flags.READONLY)` check is not relevant here right? IIUC, this checks if there's a `readonly` field in a mutable record type? Plus, isn't it the same check as in the if block above? Is the field being `readonly` and the field's type being `readonly` two different things?

MaryamZi replied:
Yeah they're different.

So there are three possible scenarios.

1. The expected type is `T & readonly` where T is the record type. Meaning, the value itself will be immutable and whatever fields that are specified should also be immutable. This is what is checked with `mutableRecordType`.

```ballerina
Foo & readonly f = {};
```

2. A field's type is `T & readonly`. This means that whatever value set for that field will have to be immutable. That's the second check.

```ballerina
type Foo record {|
    Bar & readonly b;
|};
``` 
And you can also update the `b` field in a `Foo` value after creating  the record value, but the `Bar` value must always be immutable.

3. Then there's `readonly` fields where in addition to 2) there's also the constraint that the value cannot be updated once it is set.

```ballerina
type Foo record {|
    readonly Bar b;
|};
```

We only want to log an error here for 3.
